js_modules/dagit/packages/ui/src/components/MiddleTruncate.tsx

```diff
@@ -28,88 +52,65 @@ export const MiddleTruncate: React.FC<Props> = (props) => {
     [text],
   );
 
-  React.useLayoutEffect(() => {
-    if (ref.current) {
-      const width = ref.current.getBoundingClientRect().width;
-      setTargetStyle({font: getFont(ref.current), width});
-    }
-  }, []);
-
-  const debouncedObserver = React.useMemo(() => {
-    return debounce((entry: ResizeObserverEntry) => {
-      setTargetStyle((current) => {
-        const {width} = entry.contentRect;
-        if (current) {
-          return {...current, width};
-        }
-        if (ref.current) {
-          return {font: getFont(ref.current), width};
-        }
-        return null;
-      });
-    }, 10);
-  }, []);
-
-  useResizeObserver(ref.current, debouncedObserver);
-
   return (
-    <MiddleTruncateDiv ref={ref} onCopy={handleCopy} title={text}>
-      {truncated}
-    </MiddleTruncateDiv>
+    <Container onCopy={handleCopy} title={text}>
+      <MeasureWidth ref={measure}>{text}</MeasureWidth>
+      <span>{truncatedText}</span>
+    </Container>
   );
-};
+});
 
-const MiddleTruncateDiv = styled.div`
+// An invisible target element that contains the full, no-wrapped text. This is used
+// to measure the maximum available width for our truncated string.
+const MeasureWidth = styled.div`
+  height: 0;
   overflow: hidden;
   white-space: nowrap;
 `;
 
-type TargetStyle = {
-  width: number;
-  font: string;
-};
-
-const useFixedSpan = (text: string, targetStyle: TargetStyle | null) => {
-  const [truncated, setTruncated] = React.useState(text);
-
-  React.useEffect(() => {
-    if (!targetStyle) {
-      return;
-    }
-
-    const body = document.body;
-
-    const canvas = document.createElement('canvas');
-    canvas.style.position = 'fixed';
-    canvas.style.left = '-10000px';
-    canvas.style.whiteSpace = 'nowrap';
-    canvas.style.visibility = 'hidden';
-
-    const ctx = canvas.getContext('2d');
-
-    if (!ctx) {
-      return;
-    }
-
-    const {width, font} = targetStyle;
-    if (!width) {
-      return;
-    }
-
-    const targetWidth = width;
-    ctx.font = font;
-    body.appendChild(canvas);
+const Container = styled.div`
+  overflow: hidden;
+  white-space: nowrap;
+`;
 
-    const truncated = calculateMiddleTruncation(
-      text,
-      targetWidth,
-      (value: string) => ctx.measureText(value).width,
-    );
+/**
+ * Compute the font style and maximum/necessary width for the measured container,
+ * for the specified string of text.
+ *
+ * Given those values, use a 2D canvas context to determine the longest possible
+ * middle-truncated string.
+ */
+const calculateMiddleTruncatedText = (container: HTMLDivElement, text: string) => {
```

Comment:
No longer a hook, since all state management now lives in the component instead.

Reply:
This is great! I'm a fan of this not being a hook because we could potentially use it in some DAG viz cases too where we probably want to avoid an extra render cycle for every label 🙌